### ISSUE TITLE
useLibraryCodeForTypes has no effect in Pillow

### ISSUE THREAD

**Issue by matejcik** [microsoft/pyright#7832]
**Describe the bug**
this is the reproducer code:
```python
from PIL import Image

def parse(li: list[int]):
    ...

im = Image.new(size=(100, 100), mode="L")
data = im.getdata()
reveal_type(data)
parse(data)
```
this is the pyrightconfig:
```json
{
	"useLibraryCodeForTypes": false
}
```

when Pillow is not present in the environment, the code typechecks with some warnings:
```
/home/matejcik/a/repro.py
  /home/matejcik/a/repro.py:1:6 - warning: Import "PIL" could not be resolved from source (reportMissingModuleSource)
  /home/matejcik/a/repro.py:1:17 - warning: Import "PIL.Image" could not be resolved from source (reportMissingModuleSource)
  /home/matejcik/a/repro.py:8:13 - information: Type of "data" is "Unknown"
0 errors, 2 warnings, 1 information 
```

when Pillow 10.3.0 is present, typecheck will fail:
```
/home/matejcik/a/repro.py
  /home/matejcik/a/repro.py:8:13 - information: Type of "data" is "Any | DeferredError | None"
  /home/matejcik/a/repro.py:9:7 - error: Argument of type "Any | DeferredError | None" cannot be assigned to parameter "li" of type "list[int]" in function "parse"
    Type "Any | DeferredError | None" is incompatible with type "list[int]"
      "DeferredError" is incompatible with "list[int]" (reportArgumentType)
1 error, 0 warnings, 1 information 
```

Pillow carries a `py.typed` file. The signature of `getdata()` [is this](https://github.com/python-pillow/Pillow/blob/58a47978af9f34851ce926303d05fa677010ce2a/src/PIL/Image.py#L1380C1-L1401C42):
```python
    def getdata(self, band: int | None = None):
        """(docstring left out)"""
        self.load()
        if band is not None:
            return self.im.getband(band)
        return self.im  # could be abused
```

The _inferred type_ of `self.im` is `Any | DeferredError | None`.
However, I believe that pyright has no business knowing that `self.im` will be returned when `useLibraryCodeForTypes` is false -- and that the return type of `getdata()` should be `Any` or `Unknown`, seeing as type declaration for it is missing.

**Comment by erictraut** [comment]
This is working as designed, so I don't consider it a bug.

Pyright ships with bundled stubs from typeshed, which includes stubs for Pillow. That type information is used if you don't have Pillow installed. Once you install it, the type information from the Pillow library is used instead. This is dictated by [PEP 561](https://peps.python.org/pep-0561/) which says type checkers should prefer types in a "py.typed" library over types in typeshed.

The `useLibraryCodeForTypes` has no effect on a library that has a "py.typed" marker. Once a library claims to supply type information, type checkers will attempt to use that type information from the library.

If you think that there are missing or incorrect type annotations provided by the Pillow library, please report bugs to the maintainers of that library.

I ran `pyright --verifytypes PIL` to get a sense for the "type completeness" of the library. Here's the output:
```
Symbols exported by "PIL": 2666
  With known type: 1088
  With ambiguous type: 344
  With unknown type: 1234

Type completeness score: 40.8%
```

This means the library is claiming to provide complete type information (by virtue of including a "py.typed" marker), but it is only about 40% complete.

**Comment by matejcik** [comment]
> The `useLibraryCodeForTypes` has no effect on a library that has a "py.typed" marker. Once a library claims to supply type information, type checkers will attempt to use that type information from the library.

My point of contention here is that Pillow is **not providing types** in this case.
In particular, type for `getdata()` is not provided, so pyright must have gone ahead and figured out, from code, that `self.im` is being returned. But type for `self.im` is _also not provided_, so, again, pyright must have gone ahead, parsed the library, and tried to infer the type from usage.

This, fwiw, makes it impossible for a library to be, at the same time, (1) only partially typed, and (2) provide the partial typing information inline (while (3) due to the whole thing being only partial, the inference may very well result in wrong type info).
PEP561 doesn't explicitly _allow_ partial typing information inline in package, but I would expect this situation to be relatively frequent? And by my reading that just means that the typechecker should not try to merge the partial information (step 4 MRO) with typeshed (step 5).

> This means the library is claiming to provide complete type information (by virtue of including a "py.typed" marker)

There is nothing in PEP561 saying that the type information provided needs to be "complete".

**Comment by erictraut** [comment]
Pyright used to evaluate all unannotated symbols in a "py.typed" library as `Unknown`, but we received strong feedback from library authors and consumers that they didn't like this behavior. After much discussion and negotiation, we landed on the decision to always attempt to infer types for unannotated symbols within a "py.typed" library. This was a long and drawn-out negotiation, and I'm pretty satisfied with where we ended up, so I'm not inclined to revisit that decision.

**Comment by matejcik** [comment]
alright, I didn't know that this was discussed previously. just out of curiosity, is this discussion in an issue that you could point me to?
